### PR TITLE
Issue #1565 : policy tag feature

### PR DESCRIPTION
**Pull request by sahibamittal**
Changes for new feature to support 'Limit to Tags' in policy management.
This includes:
1. endpoint to add new tag/s in a policy
2. endpoint to remove tag/s from a policy
3. endpoint to get all tags available in all projects or specific projects under a policy
4. changes in policy evaluation to limit scanning of components against policies having these tags (**if**, added in 'Limit to Tags')
5. Unit tests for above

Issue: https://github.com/DependencyTrack/dependency-track/issues/1565

Frontend changes -> https://github.com/DependencyTrack/frontend/pull/160
https://github.com/DependencyTrack/frontend/pull/162

### CODE REVIEW

File: src/main/java/org/dependencytrack/persistence/TagQueryManager.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * @param pm a PersistenceManager object
+     * @param request an AlpineRequest object
+     */
+    TagQueryManager(final PersistenceManager pm, final AlpineRequest request) {
+        super(pm, request);
+    }
+
+    public PaginatedResult getTags(String policyId) {
+
+        LOGGER.info("Retrieving tags under policy " + policyId);
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/persistence/TagQueryManager.java:41]: Please remove or change log level to `debug`.

File: src/main/java/org/dependencytrack/resources/v1/PolicyResource.java [hunk 2]
```diff
@@ -266,4 +267,82 @@ public Response removeProjectFromPolicy(
...
+    @PermissionRequired(Permissions.Constants.POLICY_MANAGEMENT)
+    public Response addTagToPolicy(
+            @ApiParam(value = "The UUID of the policy to add a project to", required = true)
+            @PathParam("policyUuid") String policyUuid,
+            @ApiParam(value = "The UUID of the tag to add to the rule", required = true)
+            @PathParam("tagUuid") String tagUuid) {
+        try (QueryManager qm = new QueryManager()) {
+            final Policy policy = qm.getObjectByUuid(Policy.class, policyUuid);
+            if (policy == null) {
+                System.out.println("The policy could not be found");
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/PolicyResource.java:293]: Please remove or use a `Logger` instance with log level `debug`.

File: src/main/java/org/dependencytrack/resources/v1/PolicyResource.java [hunk 3]
```diff
@@ -266,4 +267,82 @@ public Response removeProjectFromPolicy(
...
+                return Response.status(Response.Status.NOT_FOUND).entity("The policy could not be found.").build();
+            }
+            final Tag tag = qm.getObjectByUuid(Tag.class, tagUuid);
+            if (tag == null) {
+                System.out.println("The tag with id could not be found "+tagUuid);
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/resources/v1/PolicyResource.java:298]: Please remove or use a `Logger` instance with log level `debug`.

**Review by nscuro — CHANGES_REQUESTED** [review]
Thanks for your PR @sahibamittal! 🚀 

Few areas that could be improved or need clarification.

File: src/main/java/org/dependencytrack/resources/v1/TagResource.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            response = Tag.class,
+            responseContainer = "List",
+            responseHeaders = @ResponseHeader(name = TOTAL_COUNT_HEADER, response = Long.class, description = "The total number of tags")
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 401, message = "Unauthorized")
+    })
+    @PermissionRequired(Permissions.Constants.VIEW_PORTFOLIO)
+    public Response getTags(@ApiParam(value = "Policy Id", required = true)
+                            @QueryParam("policyId") String policyId){
```
[the remark this answers is not in the document]
Comment (sahibamittal) [reply src/main/java/org/dependencytrack/resources/v1/TagResource.java:62]: Endpoint would be to retrieve either all tags (if policy is not limited to any project) OR retrieve tags under all projects limited under the policy.
We'll not need specific `/project/{uuid}` endpoint in this case.

File: src/main/java/org/dependencytrack/persistence/TagQueryManager.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        if(projects != null && projects.size() != 0){
+            HashSet<Tag> filteredTags = new HashSet<>();
+            for(Project project : projects) {
+                filteredTags.addAll(project.getTags());
+            }
+            tagsToShow = new ArrayList<>(filteredTags);
+        } else {
+            tagsToShow = new ArrayList<>(tagsQueried);
+        }
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/persistence/TagQueryManager.java:59]: It looks like querying all tags is only necessary if no projects are associated with the given policy.

```suggestion
        List<Tag> tagsToShow;
        if(projects != null && projects.size() != 0){
            HashSet<Tag> filteredTags = new HashSet<>();
            for(Project project : projects) {
                filteredTags.addAll(project.getTags());
            }
            tagsToShow = new ArrayList<>(filteredTags);
        } else {
            tagsToShow = pm.newQuery(Tag.class).executeList();
        }
```

File: src/main/java/org/dependencytrack/persistence/TagQueryManager.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    TagQueryManager(final PersistenceManager pm, final AlpineRequest request) {
+        super(pm, request);
+    }
+
+    public PaginatedResult getTags(String policyUuid) {
+
+        LOGGER.debug("Retrieving tags under policy " + policyUuid);
+
+        QueryManager qm = new QueryManager();
+        Policy policy = qm.getObjectByUuid(Policy.class, policyUuid);
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/persistence/TagQueryManager.java:44]: Creating a new `qm` here is not necessary. You can use `this` (or omit the receiver) in the following calls.

```suggestion
        Policy policy = getObjectByUuid(Policy.class, policyUuid);
```

File: src/main/java/org/dependencytrack/model/Tag.java [hunk 7]
```diff
@@ -28,6 +29,7 @@
 import javax.jdo.annotations.Extension;
 import javax.jdo.annotations.IdGeneratorStrategy;
 import javax.jdo.annotations.Order;
+import javax.jdo.annotations.Unique;
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/model/Tag.java:32]: That import appears to be unused.

File: src/main/java/org/dependencytrack/model/Tag.java [hunk 8]
```diff
@@ -64,6 +66,7 @@ public class Tag implements Serializable {
     private String name;
 
     @Persistent
+    @JsonBackReference
```
Comment (nscuro) [inline src/main/java/org/dependencytrack/model/Tag.java:69]: `@JsonIgnore` may be a better choice here. `@JsonBackReference` has some special behavior for deserialization, for which another field annotated with `@JacksonManagedReference` is required. It doesn't appear like we ever need to deserialize the `projects` field of `Tag`.

**Review by nscuro — CHANGES_REQUESTED** [review]
Some minor nitpicks left. I tested the new functionality together with your frontend PR and it works as expected.

After the open topics have been resolved, we'll need you to sign-off your commits so your PR passes the DCO check. If you click on *Details* of the DCO check, it tells you how to do this.

File: src/main/java/org/dependencytrack/persistence/TagQueryManager.java (hunk 5 above)
Comment (sahibamittal) [reply src/main/java/org/dependencytrack/persistence/TagQueryManager.java:59]: I'll still have to use separate list for querying due to 'Query result sets are not modifiable' exception.

`else {
            List<Tag> tagsQueried = pm.newQuery(Tag.class).executeList();
            tagsToShow = new ArrayList<>(tagsQueried);
        }`

File: src/main/java/org/dependencytrack/resources/v1/TagResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sahibamittal) [reply src/main/java/org/dependencytrack/resources/v1/TagResource.java]: Yes, for projects, tags are retrieved directly under 'Projects' section. For policies, tags need to be retrieved for policy specific projects.

**Comment by nscuro** [comment]
@sahibamittal It looks like you signed off some of Steve's commits by accident.

**Comment by sahibamittal** [comment]
Closing this PR due to sign-off issue.
New PR -> https://github.com/DependencyTrack/dependency-track/pull/1694

**Pull request closed without merging**